<!DOCTYPE html>
<html class="no-js">
<head>
    <meta charset="utf-8">
    <title>Welgene NGS E-Report</title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">   
    <link href="../assets/plugins/isotope/isotope.css" rel="stylesheet" media="screen" />
    <link href="../assets/plugins/IconHoverEffects-master/css/component.css" rel="stylesheet" />
    <link href="../assets/plugins/bootstrap/bootstrap.css" rel="stylesheet" />
    <link href="../assets/css/about-achivements.css" rel="stylesheet" />
    <link id="mainStyle" href="../assets/css/style.css" rel="stylesheet" />
    <link href="../assets/css/ngsReport.css" rel="stylesheet" />
    <link href='http://fonts.googleapis.com/css?family=Varela+Round' rel='stylesheet' type='text/css'>
    <link href="../assets/plugins/fancybox_3.5.6/jquery.fancybox.css" rel="stylesheet" />
	<style>
		.whiteFont{
		 color:#FFF;
		}
		.spaceTop{
			margin-top: 25px;
		}
		.spaceBottom{
			margin-bottom: 25px;
		}
	</style>
    
<style>

.outer {
    width: 85%;
}

pre code {
	-ms-overflow-y: hidden;
	word-wrap: normal;
    overflow-wrap: normal;
    white-space: pre;
	max-width: 85%;
	min-width: 85%;
}

td code {
    white-space: normal;
}

</style>
    <script src="../assets/js/modernizr-2.6.2-respond-1.1.0.min.js"></script>
    <script src="../assets/js/jquery.js"></script>
    <script src="../assets/js/scrollReveal.js"></script>
    <script>
        window.scrollReveal = new scrollReveal(); //please put this script here to show animation at the time of scroll
    </script>
	<script src="../assets/js/jquery.easing.1.3.js"></script>
	<script src="../assets/plugins/bootstrap/bootstrap.min.js"></script>    
	<script src="../assets/plugins/isotope/jquery.isotope.min.js"></script>
	<script src="../assets/plugins/fancybox_3.5.6/jquery.fancybox.js"></script>
	<script src="../assets/js/jquery.localscroll-1.2.7-min.js"></script>
	<script src="../assets/js/jquery.appear.js"></script>
    <script src="../assets/js/velocity.min.js"></script>
</head>
<body>

    <section id="section-about" class="section" align=left>
        <div class="container" data-scrollreveal='after 0.30s'>
            <div class="row">
                <div class="col-md-offset-2 col-md-8" >
                    <div >
			
                        <h2 class="heading">Gene Fusion & Alternative Splicing</h2>

                    </div>
                </div>
            </div>
        </div>
    </section>
<div align=center>
<br>
<table width="85%" border="3">
  <tbody>
    <tr>
      <td><strong><font color="red">Gene Fusion Analysis</font></strong></td>
    </tr>
    <tr>
      <td><p><a href="https://github.com/STAR-Fusion/STAR-Fusion/" target="_blank"><strong>STAR-Fusion</strong></a>, a software that is both fast and accurate in identifying fusion transcripts from RNA-Seq data.  STAR-Fusion uses the <a href="https://github.com/alexdobin/STAR" target="_blank"><strong>STAR</strong></a> aligner to identify candidate fusion transcripts supported by Illumina reads. STAR-Fusion further processes the output generated by the STAR aligner to map junction reads and spanning reads to a reference annotation set.</p>
      <p><br>
        </p>
        <p><strong>Reference</strong><br>
        </p>
        <ol>
          <li><a href="https://www.biorxiv.org/content/10.1101/120295v1" target="_blank">Haas B, Dobin A, Stransky N, Li B, Yang X, Tickle T,Bankapur A, Ganote C, Doak T, Pochet N, Sun J, Wu C, Gingeras T, Regev A. STAR-Fusion: Fast and Accurate Fusion Transcript Detection from RNA-Seq. <em><strong>bioRxiv</strong></em> 120295; doi: https://doi.org/10.1101/120295<br>
          </a></li>
        </ol>
      </tr>
  </tbody>
</table>
<br>
<table width="85%" border="3">
  <tbody>
    <tr>
      <td width="100%"><div align="left"><strong><font color="red">Gene Fusion Results</font></strong> => <a href="Gene_Fusion/" target="_blank">Click to see the result folder</a> </div></td>
    </tr>
    <tr>
      <td>The output from STAR-Fusion is found as a tab-delimited file named '<strong>sampleName.fusion_predictions.tsv</strong>' or <strong>.xlsx</strong>, along with an abridged version that excludes the identification of the evidence fusion reads and called '<strong>sampleName.fusion_predictions.abridged.tsv</strong>' or<strong> .xlsx</strong>, with the following columns and format (demo data):</td>
    </tr>
    <tr>
      <td><code>
       <ol>
        <li><strong>FusionName</strong>: name of the genes with fusion event, represents as geneA--geneB, seperated by "--".</li>
        <li><strong>JunctionReadCount</strong>: number of RNA-Seq fragments containing a read that aligns as a split read at the site of the putative fusion junction.</li>
        <li><strong>SpanningFragCount</strong>: number of paired-end reads that span the fusion breakpoint but the reads do not directly overlap the breakpoint.</li>
        <li><strong>SpliceType</strong>: category of support at the fusion breakpoint: {
                ONLY_REF_SPLICE: fusion breakpoint occurs at reference (known) splice junctions.
                INCL_NON_REF_SPLICE: fusion breakpoint occurs at a breakpoint that does not involve all reference (known) exon junctions.
                NO_JUNCTION_READS_IDENTIFIED: only spanning fragments support the fusion. (can only happen if --min_junction_reads is set to zero)
           }</li>
        <li><strong>LeftGene</strong>: identifier of the gene represented by the left section of the fusion transcript.</li>
        <li><strong>LeftBreakpoint</strong>:position of the left fusion breakpoint in the context of the genome.</li>
        <li><strong>RightGene</strong>: identifier of the gene represented by the right section of the fusion transcript.</li>
        <li><strong>RightBreakpoint</strong>: position of the right fusion breakpoint in the context of the genome.</li>
        <li><strong>LargeAnchorSupport</strong>:  this indicates whether there are split reads that provide 'long' (set to length of 25 bases) alignments on both sides of the putative breakpoint. Those fusions supported only by split reads (no spanning fragments) and lack LargeAnchorSupport are often highly suspicious and tend to be false positives. Those with LargeAnchorSupport are labeled as 'YES_LDAS' (where LDAS = long double anchor support.... yes, more jargon).</li>
        <li><strong>FFPM</strong>: normalized measure of the quantity of RNA-Seq fragments supporting the fusion event as:  fusion fragments per total million RNA-Seq fragments. The number of fusion-supporting reads depends on both the expression of the fusion transcript and the number of reads sequenced. The deeper the sequenced data set, the greater the number of artifactual fusions that will appear with minimal supporting evidence, and so taking into account the sequencing depth is important to curtail overzealous prediction of fusion transcripts with ever-so-minimal supporting evidence. We provide normalized measures of the fusion-supporting rna-seq fragments as FFPM (fusion fragments per million total reads) measures. A filter of 0.1 sum FFPM (meaning at least 1 fusion-supporting rna-seq fragment per 10M total reads) tends to be effective at excluding fusion artifacts, and is the current default for filtering fusions from the final output.</li>
        <li><strong>LeftBreakDinuc</strong>: the genomic dinucleotides found at the left breakpoint (putative splice site if splicing is involved).</li>
        <li><strong>LeftBreakEntropy</strong>: this represent the <a href="http://bearcave.com/misl/misl_tech/wavelets/compression/shannon.html" target="_blank">Shannon entropy</a> of the 15 exonic bases flanking the breakpoint. The maximum entropy is 2, representing highest complexity. The lowest would be zero (involving a 15 base mononucleotide run). Low entropy sites should generally be treated as less confident breakpoints.</li>
        <li><strong>RightBreakDinuc</strong>: the genomic dinucleotides found at the right breakpoint (putative splice site if splicing is involved).</li>
        <li><strong>RightBreakEntropy</strong>: entropy calculation for the 15 bases immediately downstream from the fusion junction.</li>
        <li><strong>annots</strong>: The final column 'annots' provides a simplified annotation for fusion transcript, leveraging <a href="https://github.com/FusionAnnotator/FusionAnnotator/wiki" target="_blank">FusionAnnotator</a> (bundled with STAR-Fusion). </li>
       </ol>
The abridged output file contents are shown above. See the unabridged 'sampleName.fusion_predictions.tsv' output file for the identity of the RNA-Seq fragments identified as junction or spanning fragments, where the individual read names are provided as comma-delimited lists in each corresponding column ('<strong>JunctionReads</strong>' & '<strong>SpanningFrags</strong>').</code></td>
    </tr>
  </tbody>
</table>
<div class="outer">
<pre><code><strong><u>#FusionName           JunctionReadCount  SpanningFragCount  SpliceType           LeftGene                        LeftBreakpoint    RightGene                        RightBreakpoint   LargeAnchorSupport  FFPM        LeftBreakDinuc  LeftBreakEntropy  RightBreakDinuc  RightBreakEntropy  annots</u></strong>
THRA--AC090627.1      27                 93                 ONLY_REF_SPLICE      THRA^ENSG00000126351.8          chr17:38243106:+  AC090627.1^ENSG00000235300.3     chr17:46371709:+  YES_LDAS            23875.8456  GT              1.8892            AG               1.9656             ["CCLE","FA_CancerSupp","INTRACHROMOSOMAL[chr17:8.12Mb]"]
THRA--AC090627.1      5                  93                 ONLY_REF_SPLICE      THRA^ENSG00000126351.8          chr17:38243106:+  AC090627.1^ENSG00000235300.3     chr17:46384693:+  YES_LDAS            19498.6072  GT              1.8892            AG               1.4295             ["CCLE","FA_CancerSupp","INTRACHROMOSOMAL[chr17:8.12Mb]"]
ACACA--STAC2          12                 52                 ONLY_REF_SPLICE      ACACA^ENSG00000132142.15        chr17:35479453:-  STAC2^ENSG00000141750.6          chr17:37374426:-  YES_LDAS            12733.7844  GT              1.9656            AG               1.9656             ["ChimerSeq","CCLE","Klijn_CellLines","FA_CancerSupp","INTRACHROMOSOMAL[chr17:1.60Mb]"]
RPS6KB1--SNF8         10                 43                 ONLY_REF_SPLICE      RPS6KB1^ENSG00000108443.9       chr17:57970686:+  SNF8^ENSG00000159210.5           chr17:47021337:-  YES_LDAS            10545.1651  GT              1.3753            AG               1.8323             ["Klijn_CellLines","FA_CancerSupp","ChimerSeq","CCLE","INTRACHROMOSOMAL[chr17:10.95Mb]"]
TOB1--SYNRG           8                  30                 ONLY_REF_SPLICE      TOB1^ENSG00000141232.4          chr17:48943419:-  SYNRG^ENSG00000006114.11         chr17:35880751:-  YES_LDAS            7560.6844   GT              1.4566            AG               1.8892             ["FA_CancerSupp","CCLE","INTRACHROMOSOMAL[chr17:12.97Mb]"]
VAPB--IKZF3           4                  46                 ONLY_REF_SPLICE      VAPB^ENSG00000124164.11         chr20:56964573:+  IKZF3^ENSG00000161405.12         chr17:37934020:-  YES_LDAS            9948.269    GT              1.9656            AG               1.7819             ["FA_CancerSupp","Klijn_CellLines","CCLE","ChimerSeq","ChimerPub","INTERCHROMOSOMAL[chr20--chr17]"]
ZMYND8--CEP250        2                  44                 ONLY_REF_SPLICE      ZMYND8^ENSG00000101040.15       chr20:45852970:-  CEP250^ENSG00000126001.11        chr20:34078463:+  NO_LDAS             9152.4075   GT              1.8295            AG               1.8062             ["FA_CancerSupp","CCLE","ChimerSeq","INTRACHROMOSOMAL[chr20:11.74Mb]"]
AHCTF1--NAAA          3                  38                 ONLY_REF_SPLICE      AHCTF1^ENSG00000153207.10       chr1:247094880:-  NAAA^ENSG00000138744.10          chr4:76846964:-   YES_LDAS            8157.5805   GT              1.7232            AG               1.8062             ["FA_CancerSupp","CCLE","INTERCHROMOSOMAL[chr1--chr4]"]
VAPB--IKZF3           1                  46                 ONLY_REF_SPLICE      VAPB^ENSG00000124164.11         chr20:56964573:+  IKZF3^ENSG00000161405.12         chr17:37922746:-  NO_LDAS             9351.3729   GT              1.9656            AG               1.9329             ["FA_CancerSupp","Klijn_CellLines","CCLE","ChimerSeq","ChimerPub","INTERCHROMOSOMAL[chr20--chr17]"]
VAPB--IKZF3           1                  46                 ONLY_REF_SPLICE      VAPB^ENSG00000124164.11         chr20:56964573:+  IKZF3^ENSG00000161405.12         chr17:37944627:-  NO_LDAS             9351.3729   GT              1.9656            AG               1.8892             ["FA_CancerSupp","Klijn_CellLines","CCLE","ChimerSeq","ChimerPub","INTERCHROMOSOMAL[chr20--chr17]"]
</code></pre>
</div>
<BR>
<hr width="85%">
<BR>
<table width="85%" border="3">
  <tbody>
    <tr>
      <td><strong><font color="red">Alternative Splicing Analysis</font></strong></td>
    </tr>
    <tr>
      <td><p><strong><a href="https://github.com/yarden/MISO" target="_blank">MISO</a></strong> (<strong>Mixture of Isoforms</strong>) is a probabilistic framework that quantitates the expression level of alternatively spliced genes from RNA-Seq data, and identifies differentially regulated isoforms or exons across samples. By modeling the generative process by which reads are produced from isoforms in RNA-Seq, the MISO model uses Bayesian inference to compute the probability that a read originated from a particular isoform.</p>
        <p>MISO uses the inferred assignment of reads to isoforms to quantitate the abundances of the underlying set of alternative mRNA isoforms. Confidence intervals over estimates can be obtained, which quantify the reliability of the estimates.</p>
        <p><br>
        </p>
        <p><strong>Reference</strong><br>
        </p>
        <ol>
          <li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3037023/" target="_blank">Katz Y, Wang ET, Airoldi EM, Burge CB. Analysis and design of RNA sequencing
experiments for identifying isoform regulation. <em><strong>Nat Methods</strong></em>. 2010
Dec;7(12):1009-15. </a></li>
        </ol>

      </tr>
  </tbody>
</table>
<br>
<table width="85%" border="3">
  <tbody>
    <tr>
      <td width="100%"><div align="left"><strong><font color="red">Alternative Splicing  Results</font></strong> => <a href="Alternative_Splicing/" target="_blank">Click to see the result folder</a> </div></td>
    </tr>
    <tr>
      <td>The output from MISO is found as a tab-delimited file named '<strong>sampleName.misoAS.tsv</strong>' or <strong>.xlsx</strong>,  containing the following columns and format (demo data):</td>
    </tr>
    <tr>
      <td><code>
        <ol><li><strong>event_name</strong>: name of the event or gene from the input annotation.</li><li><strong>miso_posterior_mean</strong>: an estimate of Ψ (derived by taking the mean of the posterior distribution over Ψ).</li><li><strong>ci_low</strong>: the low bound of the 95% confidence interval on the Ψ estimate.</li><li><strong>ci_high</strong>: the high bound of the 95% confidence interval on the Ψ estimate.</li><li><strong>isoforms</strong>: the names of the isoforms that make up the event/gene quantitated (from the input annotations).</li><li><strong>counts</strong>: raw read counts that support each isoform. Lists the number of reads that support each combination of isoforms. Each read type is represented as a comma-separated list (enclosed in parentheses) with number of entries equal to the number of isoforms quantitated, and a 1 or 0 in each entry depending on whether the read was consistent or inconsistent with the isoform. For example, for an alternative event with two isoforms, the entry might be: (1,0):28,(0,1):50,(1,1):12, where 0 and 1 correspond to the first and second isoforms. This indicates that 28 reads support isoform 1 but not 2, while 50 reads support the second but not the first, and 12 reads support both isoforms. The field (0,0):N, if outputted, indicates the number of reads N that mapped to the region but were thrown out (e.g. because they are not consistent with the isoforms in the annotation, violate overhang constraints, or other related reasons).</li><li><strong>assigned_counts</strong>: inferred assignment of reads to isoforms. Lists each isoform (0-based) and the number of reads assigned to it. For example, for an alternative event with two isoforms, the entry might be: 0:330,1:34, indicating that 330 reads were assigned to the first isoform (isoform 0) and 34 reads were assigned to the second isoform (isoform 1).</li><li><strong>chrom</strong>: chromosome name.</li><li><strong>strand</strong>: strand of gene/event.</li><li><strong>mRNA_starts</strong>: comma-separated list of start genomic coordinates for mRNA entries in gene/event.</li><li><strong>mRNA_ends</strong>: comma-separated list of end genomic coordinates for mRNA enries in gene/event.</li>
     </ol>
When one of those fields is not available (e.g. <strong>strand</strong>), NA is outputted in its place.<BR>

Note that <strong>miso_posterior_mean</strong> is an estimate of Ψ and [ci_low, ci_high] are the 95% confidence limits of that estimate.
        </code></td>
    </tr>
  </tbody>
</table>
<div class="outer">
<pre><code><strong><u>event_name              miso_posterior_mean     ci_low                  ci_high                 isoforms                                                                                                        counts                                                                                                          assigned_counts         chrom	strand	mRNA_starts	                        mRNA_ends</u></strong>
gene:ENSG00000108771	0.19,0.20,0.30,0.31	0.09,0.06,0.07,0.04	0.35,0.41,0.59,0.57	'transcript:ENST00000251642@42101404@42101946@-_transcript:ENST00000251642@42102216@42102312@-(...omit)'	(0,0,0,0):12,(0,1,0,0):3,(0,1,0,1):1,(1,0,0,0):17,(1,0,1,0):1,(1,1,0,0):5,(1,1,0,1):7,(1,1,1,0):1,(1,1,1,1):4	0:20,1:12,2:1,3:6	17	-	42101404,42107693,42110779,42110816	42112733,42112486,42112714,42112463
gene:ENSG00000136451	0.56,0.03,0.34,0.07	0.48,0.00,0.25,0.02	0.63,0.10,0.43,0.19	'transcript:ENST00000258963@57971547@57974900@-_transcript:ENST00000258963@57979152@57979313@-(...omit)'	(0,0,0,0):41,(0,0,1,0):7,(0,1,1,0):30,(0,1,1,1):21,(1,0,0,0):242,(1,1,1,0):158	                                0:340,1:22,2:92,3:4	17	-	57971547,57973759,57973759,57982910	57982864,57985348,57988152,57988259
gene:ENSG00000141503	0.54,0.08,0.29,0.09	0.28,0.00,0.02,0.01	0.86,0.24,0.55,0.27	'transcript:ENST00000355280@4833388@4833640@+_transcript:ENST00000355280@4878317@4878382@+_(...omit)'	        (0,0,0,0):30,(1,0,0,0):2,(1,0,1,0):9,(1,1,0,0):17,(1,1,1,0):91,(1,1,1,1):3	                                0:63,1:10,2:49	        17	+	4833388,4833400,4833584,4896205	        4898053,4898054,4897287,4897145
gene:ENSG00000276023	0.42,0.50,0.08	        0.02,0.03,0.00	        0.93,0.94,0.28	        'transcript:ENST00000617516@37489831@37489958@+_transcript:ENST00000617516@37510677@37510764@+_(...omit)'	(0,0,0):1,(1,1,1):21	                                                                                        0:17,1:1,2:3	        17	+       37489831,37491464,37509879	        37513501,37513487,37513501
</code></pre></div>
<br>
</div>
</body>
</html>
